File: ios/native_workmanager/Sources/native_workmanager/NativeWorkmanagerPlugin+RemoteTrigger.swift
import Foundation
import KMPWorkManager
import Flutter
import CryptoKit

extension NativeWorkmanagerPlugin {

    internal func handleRegisterRemoteTrigger(call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let args = call.arguments as? [String: Any],
              let source = args["source"] as? String,
              let ruleMap = args["rule"] as? [String: Any],
              let payloadKey = ruleMap["payloadKey"] as? String,
              let workerMappings = ruleMap["workerMappings"] as? [String: Any] else {
            result(FlutterError(code: "INVALID_ARGS", message: "Missing required arguments", details: nil))
            return
        }

        guard let mappingsData = try? JSONSerialization.data(withJSONObject: workerMappings),
              let mappingsJson = String(data: mappingsData, encoding: .utf8) else {
            result(FlutterError(code: "SERIALIZATION_ERROR", message: "Failed to serialize worker mappings", details: nil))
            return
        }

        let secretKey = ruleMap["secretKey"] as? String

        if #available(iOS 13.0, *) {
            RemoteTriggerStore.shared.upsert(
                source: source,
                payloadKey: payloadKey,
                workerMappingsJson: mappingsJson,
                secretKey: secretKey
            )
        }

        NativeLogger.d("✅ Remote trigger registered for \(source) (key: \(payloadKey), hmac: \(secretKey != nil))")
        result(nil)
    }

    /// Handle a remote notification and optionally trigger a native worker.
    @objc public static func onRemoteNotification(userInfo: [AnyHashable: Any],
                                                 completionHandler: @escaping (UIBackgroundFetchResult) -> Void) -> Bool {
        guard #available(iOS 13.0, *) else {
            completionHandler(.noData)
            return false
        }

        let payload = userInfo as? [String: Any] ?? [:]
        let source = "fcm"
        
        guard let record = RemoteTriggerStore.shared.getRule(source: source) else {
            if let command = payload["native_wm"] {
                NativeLogger.d("📡 Processing direct command (Native WM)")
                Task {
                    let cmdMap = (command as? [String: Any]) ?? (try? JSONSerialization.jsonObject(with: (command as? String ?? "").data(using: .utf8) ?? Data()) as? [String: Any]) ?? [:]
                    let success = await CommandProcessor.handleDirectRemoteCommand(command: cmdMap)
                    completionHandler(success ? .newData : .failed)
                }
                return true
            }
            completionHandler(.noData)
            return false
        }

        // Phase 2: HMAC Verification
        if let secretKey = record.secretKey {
            let signature = payload["x-native-wm-signature"] as? String
            if signature == nil || !verifyHmac(payload: payload, secretKey: secretKey, signature: signature!) {
                NativeLogger.w("⚠️ Remote trigger REJECTED: Invalid HMAC signature for \(source)")
                completionHandler(.failed)
                return false
            }
        }

        // Mode 1: Direct Command (Highest priority)
        if let command = payload["native_wm"] {
            NativeLogger.d("📡 Processing direct command (Native WM)")
            Task {
                let cmdMap = (command as? [String: Any]) ?? (try? JSONSerialization.jsonObject(with: (command as? String ?? "").data(using: .utf8) ?? Data()) as? [String: Any]) ?? [:]
                let success = await CommandProcessor.handleDirectRemoteCommand(command: cmdMap)
                completionHandler(success ? .newData : .failed)
            }
            return true
        }

        // Mode 2: Rule-based Mapping
        var triggerValue: String?
        if let val = payload[record.payloadKey] {
            triggerValue = "\(val)"
        } else if let data = payload["data"] as? [String: Any], let val = data[record.payloadKey] {
            triggerValue = "\(val)"
        }

        guard let val = triggerValue else {
            completionHandler(.noData)
            return false
        }

        NativeLogger.d("📡 Processing remote trigger: \(source) (mapped via \(record.payloadKey))")

        guard let mappingsData = record.workerMappingsJson.data(using: .utf8),
              let mappings = try? JSONSerialization.jsonObject(with: mappingsData) as? [String: Any],
              let mapping = mappings[val] as? [String: Any],
              let workerClassName = mapping["workerClassName"] as? String,
              let workerConfig = mapping["workerConfig"] as? [String: Any] else {
            completionHandler(.noData)
            return false
        }

        let substitutedConfig = CommandProcessor.substituteTemplates(in: workerConfig, with: payload)
        let taskId = "remote_\(val)_\(UUID().uuidString.prefix(8))"

        NativeLogger.d("✅ Remote trigger matched '\(val)': Executing \(workerClassName) (\(taskId))")

        Task {
            let success = await CommandProcessor.executeWorkerStateless(
                taskId: taskId,
                workerClassName: workerClassName,
                workerConfig: substitutedConfig
            )
            completionHandler(success ? .newData : .failed)
        }

        return true
    }

    private static func verifyHmac(payload: [String: Any], secretKey: String, signature: String) -> Bool {
        if #available(iOS 13.0, *) {
            let filteredKeys = payload.keys.filter { $0 != "x-native-wm-signature" }.sorted()
            let dataToSign = filteredKeys.map { key in
                let value = payload[key]
                let valueStr: String
                
                if let dict = value as? [String: Any] {
                    let data = try? JSONSerialization.data(withJSONObject: dict, options: [.sortedKeys, .fragmentsAllowed])
                    valueStr = data.flatMap { String(data: $0, encoding: .utf8) } ?? "\(value ?? "null")"
                } else if let array = value as? [Any] {
                    let data = try? JSONSerialization.data(withJSONObject: array, options: [.sortedKeys, .fragmentsAllowed])
                    valueStr = data.flatMap { String(data: $0, encoding: .utf8) } ?? "\(value ?? "null")"
                } else {
                    valueStr = "\(value ?? "null")"
                }
                return "\(key)=\(valueStr)"
            }.joined(separator: "|")
            
            guard let keyData = secretKey.data(using: .utf8),
                  let messageData = dataToSign.data(using: .utf8) else {
                return false
            }
            
            let key = SymmetricKey(data: keyData)

            // Constant-time verification. Swift's String `==` short-circuits on the
            // first mismatch, which can leak how many leading bytes matched (timing
            // side-channel). CryptoKit's isValidAuthenticationCode compares in
            // constant time. Decode the provided hex first so we compare raw bytes.
            guard let providedMac = Self.dataFromHexString(signature) else { return false }
            return HMAC<SHA256>.isValidAuthenticationCode(
                providedMac, authenticating: messageData, using: key)
        }
        return false
    }

    /// Decode a hex string to Data, or nil if it is malformed (odd length / non-hex).
    private static func dataFromHexString(_ hex: String) -> Data? {
        guard !hex.isEmpty, hex.count % 2 == 0 else { return nil }
        var data = Data(capacity: hex.count / 2)
        var index = hex.startIndex
        while index < hex.endIndex {
            let next = hex.index(index, offsetBy: 2)
            guard let byte = UInt8(hex[index..<next], radix: 16) else { return nil }
            data.append(byte)
            index = next
        }
        return data
    }
}

File: android/src/main/kotlin/dev/brewkits/native_workmanager/NativeWorkmanagerPlugin+RemoteTrigger.kt
package dev.brewkits.native_workmanager

import android.content.Context
import dev.brewkits.native_workmanager.store.RemoteTriggerStore
import dev.brewkits.native_workmanager.utils.CommandProcessor
import dev.brewkits.native_workmanager.utils.MappingUtils
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel.Result
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import org.json.JSONArray
import org.json.JSONObject
import java.util.UUID

internal fun NativeWorkmanagerPlugin.handleRegisterRemoteTrigger(call: MethodCall, result: Result) {
    scope.launch {
        try {
            val source = call.argument<String>("source")
                ?: return@launch result.error("INVALID_ARGS", "source required", null)
            val ruleMap = call.argument<Map<String, Any?>>("rule")
                ?: return@launch result.error("INVALID_ARGS", "rule required", null)

            val payloadKey = ruleMap["payloadKey"] as? String
                ?: return@launch result.error("INVALID_ARGS", "payloadKey required", null)
            val workerMappings = ruleMap["workerMappings"] as? Map<String, Any?>
                ?: return@launch result.error("INVALID_ARGS", "workerMappings required", null)

            val mappingsJson = MappingUtils.toJson(workerMappings)
            val secretKey = ruleMap["secretKey"] as? String

            withContext(Dispatchers.IO) {
                remoteTriggerStore.upsert(
                    source = source,
                    payloadKey = payloadKey,
                    workerMappingsJson = mappingsJson,
                    secretKey = secretKey
                )
            }

            NativeLogger.d("✅ Remote trigger registered for $source (key: $payloadKey, hmac: ${secretKey != null})")
            result.success(null)
        } catch (e: Exception) {
            NativeLogger.e("❌ Register remote trigger error", e)
            result.error("REGISTER_REMOTE_TRIGGER_ERROR", e.message, null)
        }
    }
}

/**
 * Handle a remote message (FCM/APNs) and optionally trigger a native worker.
 */
fun NativeWorkmanagerPlugin.Companion.onRemoteMessage(context: Context, source: String, payload: Map<String, Any?>): Boolean {
    try {
        val store = RemoteTriggerStore(context)
        val record = store.getRule(source) ?: return false

        // Phase 2: HMAC Verification
        if (record.secretKey != null) {
            val signature = payload["x-native-wm-signature"]?.toString()
            if (signature == null || !verifyHmac(payload, record.secretKey, signature)) {
                NativeLogger.w("⚠️ Remote trigger REJECTED: Invalid HMAC signature for $source")
                return false
            }
        }

        // Mode 1: Direct Command (Highest priority)
        val nativeWmCommand = payload["native_wm"] ?: (payload["data"] as? Map<*, *>)?.get("native_wm")
        
        if (nativeWmCommand != null) {
            NativeLogger.d("📡 Processing direct remote command (Native WM)")
            if (nativeWmCommand is Map<*, *>) {
                @Suppress("UNCHECKED_CAST")
                if (CommandProcessor.handleDirectRemoteCommand(context, nativeWmCommand as Map<String, Any?>)) {
                    return true
                }
            } else if (nativeWmCommand is String) {
                try {
                    val json = JSONObject(nativeWmCommand)
                    if (CommandProcessor.handleDirectRemoteCommand(context, CommandProcessor.jsonToMap(json))) {
                        return true
                    }
                } catch (_: Exception) {}
            }
        }

        // Mode 2: Rule-based Mapping
        val triggerValue = payload[record.payloadKey]?.toString() 
            ?: (payload["data"] as? Map<*, *>)?.get(record.payloadKey)?.toString()
            ?: return false
            
        NativeLogger.d("📡 Processing remote trigger: $source (mapped via ${record.payloadKey})")

        val mappings = JSONObject(record.workerMappingsJson)
        if (!mappings.has(triggerValue)) return false

        val mapping = mappings.getJSONObject(triggerValue)
        val workerClassName = mapping.getString("workerClassName")
        val workerConfigJson = mapping.optString("workerConfig", "{}")

        // Perform template substitution
        val substitutedConfig = try {
            val json = JSONObject(workerConfigJson)
            substituteInJsonObject(json, payload)
            json.toString()
        } catch (e: Exception) {
            NativeLogger.e("❌ Error substituting templates in workerConfig", e)
            workerConfigJson
        }

        val taskId = "remote_${triggerValue}_${UUID.randomUUID().toString().take(8)}"
        
        CommandProcessor.enqueueFromRemote(
            context = context,
            taskId = taskId,
            workerClassName = workerClassName,
            inputJson = substitutedConfig
        )

        NativeLogger.d("✅ Remote trigger matched '$triggerValue': Enqueued $workerClassName ($taskId)")
        return true
    } catch (e: Exception) {
        NativeLogger.e("❌ Error handling remote message", e)
        return false
    }
}

private fun verifyHmac(payload: Map<String, Any?>, secretKey: String, signature: String): Boolean {
    try {
        val dataToSign = payload.toMutableMap().apply { remove("x-native-wm-signature") }
            .toSortedMap()
            .entries.joinToString("|") { entry ->
                val valueStr = when (val v = entry.value) {
                    null -> "null"
                    is Map<*, *>, is List<*> -> MappingUtils.toJson(v)
                    else -> v.toString()
                }
                "${entry.key}=$valueStr"
            }

        val keySpec = javax.crypto.spec.SecretKeySpec(secretKey.toByteArray(), "HmacSHA256")
        val mac = javax.crypto.Mac.getInstance("HmacSHA256")
        mac.init(keySpec)
        val hmacBytes = mac.doFinal(dataToSign.toByteArray())

        // Constant-time comparison to avoid a timing side-channel on signature
        // verification. String.equals short-circuits on the first mismatch, which
        // can leak how many leading bytes matched. MessageDigest.isEqual compares
        // in constant time. Decode the provided hex first so we compare raw bytes.
        val providedBytes = hexStringToBytes(signature) ?: return false
        return java.security.MessageDigest.isEqual(hmacBytes, providedBytes)
    } catch (e: Exception) {
        NativeLogger.e("HMAC Verification failed", e)
        return false
    }
}

/** Decode a hex string to bytes, or null if it is malformed (odd length / non-hex). */
private fun hexStringToBytes(hex: String): ByteArray? {
    if (hex.isEmpty() || hex.length % 2 != 0) return null
    return try {
        ByteArray(hex.length / 2) { i ->
            val hi = Character.digit(hex[i * 2], 16)
            val lo = Character.digit(hex[i * 2 + 1], 16)
            if (hi < 0 || lo < 0) throw NumberFormatException("non-hex character")
            ((hi shl 4) or lo).toByte()
        }
    } catch (e: Exception) {
        null
    }
}

private fun substituteInJsonObject(json: JSONObject, values: Map<String, Any?>) {
    val keys = json.keys()
    while (keys.hasNext()) {
        val key = keys.next()
        when (val value = json.get(key)) {
            is String -> {
                if (value.contains("{{") && value.contains("}}")) {
                    json.put(key, substituteString(value, values))
                }
            }
            is JSONObject -> substituteInJsonObject(value, values)
            is JSONArray -> {
                for (i in 0 until value.length()) {
                    val item = value.get(i)
                    if (item is JSONObject) {
                        substituteInJsonObject(item, values)
                    } else if (item is String) {
                        if (item.contains("{{") && item.contains("}}")) {
                            value.put(i, substituteString(item, values))
                        }
                    }
                }
            }
        }
    }
}

private fun substituteString(template: String, values: Map<String, Any?>): String {
    var result = template
    values.forEach { (key, value) ->
        val placeholder = "{{$key}}"
        if (result.contains(placeholder)) {
            result = result.replace(placeholder, value?.toString() ?: "null")
        }
    }
    (values["data"] as? Map<*, *>)?.forEach { (key, value) ->
        val placeholder = "{{$key}}"
        if (result.contains(placeholder)) {
            result = result.replace(placeholder, value?.toString() ?: "null")
        }
    }
    return result
}
